File: iacquabot/scripts/Basic_system/world2body.py
#!/usr/bin/env python3

import rospy
import numpy as np
from nav_msgs.msg import Odometry

class OdometerNode:
    def __init__(self):
        # Inicializa el nodo
        rospy.init_node('world2body_node', anonymous=True)

        # Suscripción al tópico de odometría
        self.subscriber = rospy.Subscriber('/wamv/sensors/position/p3d_wamv', Odometry, self.odometry_callback)

        # Publicador para el nuevo tópico de odometría
        self.odom_pub = rospy.Publisher("boat/odom", Odometry, queue_size=10)

        # Tiempo de publicación a 100 Hz
        self.rate = rospy.Rate(100)  # 100 Hz

    def odometry_callback(self, msg):
        # Toma los datos de la odometría
        odom_data = msg

        # Extrae la posición
        position = odom_data.pose.pose.position
        orientation = odom_data.pose.pose.orientation

        # Extrae las velocidades
        linear_velocity = odom_data.twist.twist.linear
        angular_velocity = odom_data.twist.twist.angular

        # Convertir orientación de cuaterniones a ángulo de Euler (yaw)
        _, _, psi = self.quaternion_to_euler(orientation)

        # Cálculo de la matriz de rotación
        J = np.array([[np.cos(psi), -np.sin(psi), 0],
                      [np.sin(psi),  np.cos(psi), 0],
                      [0, 0, 1]])

        # Crear el vector de velocidades en el marco inercial
        dot_eta = np.array([linear_velocity.x, linear_velocity.y, angular_velocity.z])

        # Calcular las velocidades en el marco de cuerpo
        nu = np.linalg.inv(J).dot(dot_eta)

        # Crear nuevo mensaje de odometría
        new_odom = Odometry()
        new_odom.header.stamp = rospy.Time.now()
        new_odom.header.frame_id = "odom"
        new_odom.child_frame_id = "base_link"

        # Asigna la nueva posición y orientación
        new_odom.pose.pose.position = position
        new_odom.pose.pose.orientation = orientation

        # Asigna las velocidades en el marco de cuerpo
        new_odom.twist.twist.linear.x = nu[0]  # Velocidad en X (cuerpo)
        new_odom.twist.twist.linear.y = nu[1]  # Velocidad en Y (cuerpo)
        new_odom.twist.twist.angular.z = nu[2]  # Velocidad angular (cuerpo)

        # Publica la nueva odometría
        self.odom_pub.publish(new_odom)

    def quaternion_to_euler(self, orientation):
        """
        Convierte cuaterniones a ángulos de Euler (yaw).
        """
        # Extraer los componentes del cuaternión
        x = orientation.x
        y = orientation.y
        z = orientation.z
        w = orientation.w
        
        # Calcular yaw (psi)
        yaw = np.arctan2(2.0 * (w * z + x * y), 1.0 - 2.0 * (y * y + z * z))
        return (0.0, 0.0, yaw)  # Retorna solo el yaw

    def run(self):
        # Ciclo de ejecución del nodo
        while not rospy.is_shutdown():
            self.rate.sleep()  # Mantiene la tasa de 100 Hz

if __name__ == '__main__':
    try:
        odometer_node = OdometerNode()
        odometer_node.run()
    except rospy.ROSInterruptException:
        pass

File: iacquabot/src/world2body.cpp
#include <ros/ros.h>
#include <nav_msgs/Odometry.h>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <cmath>

class OdometerNode {
public:
    OdometerNode() {
        ros::NodeHandle nh;

        odom_sub_ = nh.subscribe("/iacquabot/sensors/position/p3d_wamv", 10, &OdometerNode::odometryCallback, this);
        odom_pub_ = nh.advertise<nav_msgs::Odometry>("boat/odom", 10);
    }

    void spin() {
        ros::Rate rate(100);  // 100 Hz
        while (ros::ok()) {
            ros::spinOnce();
            rate.sleep();
        }
    }

private:
    ros::Subscriber odom_sub_;
    ros::Publisher odom_pub_;

    void odometryCallback(const nav_msgs::Odometry::ConstPtr& msg) {
        nav_msgs::Odometry new_odom;

        // Extraer posición y orientación
        geometry_msgs::Point position = msg->pose.pose.position;
        geometry_msgs::Quaternion orientation = msg->pose.pose.orientation;

        // Extraer velocidades lineales y angulares
        geometry_msgs::Vector3 linear_velocity = msg->twist.twist.linear;
        geometry_msgs::Vector3 angular_velocity = msg->twist.twist.angular;

        // Convertir cuaternión a yaw
        tf2::Quaternion q(orientation.x, orientation.y, orientation.z, orientation.w);
        double roll, pitch, yaw;
        tf2::Matrix3x3(q).getRPY(roll, pitch, yaw);

        // Construir matriz de rotación (solo en 2D, yaw)
        double c = std::cos(yaw);
        double s = std::sin(yaw);
        double inv_det = 1.0 / (c * c + s * s);  // Por seguridad (aunque en teoría siempre es 1)

        // Transformar velocidades del marco inercial al marco cuerpo
        double vx_inertial = linear_velocity.x;
        double vy_inertial = linear_velocity.y;
        double wz_inertial = angular_velocity.z;

        double vx_body =  inv_det * ( c * vx_inertial + s * vy_inertial );
        double vy_body =  inv_det * (-s * vx_inertial + c * vy_inertial );
        double wz_body = wz_inertial;

        // Preparar nuevo mensaje de odometría
        new_odom.header.stamp = ros::Time::now();
        new_odom.header.frame_id = "odom";
        new_odom.child_frame_id = "base_link";

        new_odom.pose.pose.position = position;
        new_odom.pose.pose.orientation = orientation;

        new_odom.twist.twist.linear.x = vx_body;
        new_odom.twist.twist.linear.y = vy_body;
        new_odom.twist.twist.angular.z = wz_body;

        // Publicar mensaje
        odom_pub_.publish(new_odom);
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "world2body_node");
    OdometerNode node;
    node.spin();
    return 0;
}
